pymodelextractor/learners/observation_table_learners/generic_lstar_learner.py

```diff
@@ -0,0 +1,239 @@
+            counter += 1
+
+        result = self._learning_results_for(model, time.time() - start_time)
+        duration = time.time() - start_time
+        if self.log_hierachy != no_log:
+            print("**** Learning finished in " + str(duration) + "s using " + str(counterexample_counter) \
+                + " counterexamples & final model ended with " + str(result.state_count) + " states ****" + '\n')
+        if self._max_time != None: return_dict[self._max_time] = result
+        return result
+
+    def _update_observation_table_with(self, counterexample) -> bool:
+        prefixes = counterexample.get_prefixes()
+        for sequence in prefixes:
+            surpassed_max_query_len = self._add_to_red(sequence)
+            if surpassed_max_query_len:
+                        return surpassed_max_query_len
+            
+            for symbol in self._symbols:
+                suffixedSequence = sequence + symbol
+                if suffixedSequence not in prefixes:
+                    surpassed_max_query_len = self._add_to_blue(suffixedSequence)
+
+                    if surpassed_max_query_len:
+                        return surpassed_max_query_len
+        return False
+    
+    def _learning_results_for(self, model, duration):
+        number_of_states = len(model.states) if model is not None else 0
+        info = {
+            'equivalence_queries_count': self._teacher.equivalence_queries_count,
+            'membership_queries_count': self._teacher.membership_queries_count,
+            'observation_table': self._observation_table,
+            'duration': duration,
+        }
+        return LearningResult(model, number_of_states, info)
+
+    def _close(self) -> bool:
+        start_closing_time = time.time()
+        while True:
+            closed_counter_example = self._observation_table.is_closed()
+            if closed_counter_example == None:
+                duration = time.time() - start_closing_time
+                if self.log_hierachy == trace_log:
+                    print("    . Closed table in " + str(duration) + "s")
+                return False
+            self._observation_table.move_from_blue_to_red(closed_counter_example)
+            surpassed_max_query_len = self._add_suffixes_to_blue(closed_counter_example)
+
+            if surpassed_max_query_len:
+                return surpassed_max_query_len
+    
+
+    def _add_suffixes_to_blue(self, sequence: Sequence) -> bool:
+        for symbol in self._symbols:
+            surpassed_max_query_len = self._add_to_blue(sequence + symbol)
+            if surpassed_max_query_len:
+                return surpassed_max_query_len
+            
+        return False
+
+    def _make_consistent(self) -> bool:
+        while True:
+            start_consistent_time = time.time()
+            inconsistency = self._observation_table.find_inconsistency(self._teacher.alphabet)
+            duration = time.time() - start_consistent_time
```

Comment:
should we always allocate the memory for time if it's not used? idk, ignore this for now